# UTS Text Mining
# Nama - Kelas - NIM : Nicholas Javier - LA05 - 2502041956
# Video Penjelasan: https://youtu.be/Z1spbXNlIXA
# Backup Video: https://youtu.be/Z1spbXNlIXA

In [ ]:
!pip install newspaper3k

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.1/211.1 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.3/81.3 kB 10.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.6/97.6 kB 8.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 73.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Created wheel for tinysegmenter: filename=tinysegmenter-0.3-py3-none-any.whl size=13538 sha256=65abc596ed85aa27aac3a62bc69ba8879db457f30ac70daf19bd09e05fa8d029
  Stored in directory: /root/.cache/pip/wheels/c8/d6/6c/384f58df48c00b9a31d638005143b5b3ac62c3d25fb1447f23
  Created wheel for feedfinder2: filename=feedfinder2-0.0.4-py3-none-any.whl size=3340 sha256=9e958a31a2bed028a0584d4ed84f00b2384ed0326a03a58f162cef0ede488d4c
  Stored in directory: /root/.cache/pip/wheels/97/02/e7/a1ff1760e12bdbaab0ac824fae5c1b

Pertama-tama kita langsung melakukan !pip install newspaper3k untuk mendapatkan berita dan artikel yang di website online. DIgunakan untuk scrapping dan digunakan untuk menerima beberapa tulisan informasi yang penting dari berita tersebut

## 1. Scrapping

In [ ]:
import newspaper
from newspaper import build
import pandas as pd

sources = [
    ('https://news.kompas.com/', 'politik'), #Kompas Politik
    ('https://www.antaranews.com/politik', 'politik'), # Antarnews Politik
    ('https://pemilu.tempo.co/','politik'), # Tempo Politik
    ('https://sports.okezone.com/', 'olahraga'),# Okezone Politik
    ('https://bola.kompas.com/', 'olahraga'),# Olahraga Kompas
    ('https://www.metrotvnews.com/channel/olahraga', 'olahraga'),  # Metro Olahraga
    ('https://money.kompas.com/', 'bisnis'), # Kompas Bisnis
    ('https://www.cnbcindonesia.com/', 'bisnis'),
    ('https://www.suara.com/entertainment/entertainment-category/gosip', 'hiburan'),
    ('https://entertainment.kompas.com/', 'hiburan') # Kompas Entertainment
]

def scrape_articles(sources, count):
    articles = []
    for url, category in sources:
        paper = build(url, memoize_articles=False, language='id')
        counter = 0
        for article in paper.articles[1:count + 1]:
            try:
                article.download()
                article.parse()
                articles.append({
                    'text': article.text,
                    'media': paper.brand,
                    'label': category
                })
                counter += 1
            except Exception as e:
                print('Not Found')
    return articles

newscraper = scrape_articles(sources, 12)
df = pd.DataFrame(newscraper)
print(df)

CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://news.kompas.com/rss
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://news.kompas.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://news.kompas.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.antaranews.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] HTTPSConnectionPool(host='korporat.tempo.co', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x78c96dc8e860>, 'Connection to korporat.tempo.co timed out. (connect timeout=7)'))
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://pemilu.tempo.co/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://pemilu.tempo.co/feed
CRITICAL:newspaper.network:[REQUEST FAIL

                                                  text   media    label
0    JAKARTA, KOMPAS.com - Pemilihan Kepala Daerah ...  kompas  politik
1    JAKARTA, KOMPAS.com - Polisi masih memeriksa k...  kompas  politik
2    JAKARTA, KOMPAS.com - Agusmita (27) kini menja...  kompas  politik
3    \n\nJAKARTA, KOMPAS.com - Jajaran Satuan Reser...  kompas  politik
4    \n\nYOGYAKARTA, KOMPAS.com - Muhammadiyah meng...  kompas  politik
..                                                 ...     ...      ...
115  KOMPAS.com - Penggemar anime sangat menantikan...  kompas  hiburan
116  JAKARTA, KOMPAS.com - Komedi Kacau merupakan s...  kompas  hiburan
117  JAKARTA, KOMPAS.com - Reclaim merupakan sebuah...  kompas  hiburan
118  JAKARTA, KOMPAS.com - Film action thriller Tor...  kompas  hiburan
119  JAKARTA, KOMPAS.com - Catheez merupakan gamer ...  kompas  hiburan

[120 rows x 3 columns]


Cell code diatas merupakan bagian dari scrapping dimana saya awalnya memasukan beberapa URL atau link yang saya simpan di variable sources. Setelah itu di download, di parse, mengambil tulisan dengan adanya context. Setelah itu saya mengambil dari media, category, dan isi dari teksnya. Kemudian saya simpas di dataframe dan kemudian di print.

In [ ]:
df.to_csv('UTS_CSV.csv',index = False)

Saya menyimpan dataframe yang tadi kedalam UTS_CSV.csv

In [ ]:
df

text   media    label
0    JAKARTA, KOMPAS.com - Pemilihan Kepala Daerah ...  kompas  politik
1    JAKARTA, KOMPAS.com - Polisi masih memeriksa k...  kompas  politik
2    JAKARTA, KOMPAS.com - Agusmita (27) kini menja...  kompas  politik
3    \n\nJAKARTA, KOMPAS.com - Jajaran Satuan Reser...  kompas  politik
4    \n\nYOGYAKARTA, KOMPAS.com - Muhammadiyah meng...  kompas  politik
..                                                 ...     ...      ...
115  KOMPAS.com - Penggemar anime sangat menantikan...  kompas  hiburan
116  JAKARTA, KOMPAS.com - Komedi Kacau merupakan s...  kompas  hiburan
117  JAKARTA, KOMPAS.com - Reclaim merupakan sebuah...  kompas  hiburan
118  JAKARTA, KOMPAS.com - Film action thriller Tor...  kompas  hiburan
119  JAKARTA, KOMPAS.com - Catheez merupakan gamer ...  kompas  hiburan

[120 rows x 3 columns]

Saya hanya mengeprint lagi untuk memeriksa apakah ada yang duplikat atau isi teks, media, category sudah bener.

##2. Melakukan text preprocessing seperti cleansing, tokenization, filtering dan anda dapat menambahkan stemming atau lemmatization jika diperlukan. Sediakan hasil cleansing dalam 1 kolom terpisah sehingga anda memiliki table sebagai berikut:


Berikutnya yang saya lakukan adalah bagian preprocessing

In [ ]:
df = pd.read_csv('UTS_CSV.csv')

Agar sudah tidak perlu run yang atas karena di runnya cukup lama. Saya hanya perlu mengimpor UTS_CSV.csv agar tidak perlu run lagi

In [ ]:
df.head(5)

text   media    label
0  JAKARTA, KOMPAS.com - Pemilihan Kepala Daerah ...  kompas  politik
1  JAKARTA, KOMPAS.com - Polisi masih memeriksa k...  kompas  politik
2  JAKARTA, KOMPAS.com - Agusmita (27) kini menja...  kompas  politik
3  JAKARTA, KOMPAS.com - Presiden Partai Keadilan...  kompas  politik
4  KOMPAS.com - Presiden Joko Widodo (Jokowi) dij...  kompas  politik

Hanya melakukan pengecekan untuk data yang sudah ada

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117 entries, 0 to 116
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    117 non-null    object
 1   media   117 non-null    object
 2   label   117 non-null    object
dtypes: object(3)
memory usage: 2.9+ KB


Untuk mencari info dataframe

In [ ]:
df.describe()

text   media     label
count                                                 117     117       117
unique                                                113       7         4
top     SWEDISH Ambassador to Indonesia, Marina Berg, ...  kompas  olahraga
freq                                                    3      46        36

Hanya mencari description dari beberapa kolom

In [ ]:
df.isna().sum()

text     0
media    0
label    0
dtype: int64

Mencari data kosong dan hasilnya tidak ada NULL data

In [ ]:
import pandas as pd
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import string

nltk.download('stopwords')
nltk.download('punkt')

def preprocess_text(text):

    tokens = word_tokenize(text.lower())
    stop_words = set(stopwords.words('indonesian'))
    punctuation = set(string.punctuation)
    filtered_tokens = [word for word in tokens if word.isalnum() and word not in stop_words and word not in punctuation]
    clean_text = ' '.join(filtered_tokens)

    return clean_text


df['clean_text'] = df['text'].apply(preprocess_text)
df_preprocess = df[['text', 'clean_text','media','label']]
df_preprocess

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


text  \
0    JAKARTA, KOMPAS.com - Pemilihan Kepala Daerah ...   
1    JAKARTA, KOMPAS.com - Polisi masih memeriksa k...   
2    JAKARTA, KOMPAS.com - Agusmita (27) kini menja...   
3    JAKARTA, KOMPAS.com - Presiden Partai Keadilan...   
4    KOMPAS.com - Presiden Joko Widodo (Jokowi) dij...   
..                                                 ...   
112  KOMPAS.com - Penggemar anime sangat menantikan...   
113  JAKARTA, KOMPAS.com - Komedi Kacau merupakan s...   
114  JAKARTA, KOMPAS.com - Reclaim merupakan sebuah...   
115  JAKARTA, KOMPAS.com - Film action thriller Tor...   
116  JAKARTA, KOMPAS.com - Catheez merupakan gamer ...   

                                            clean_text   media    label  
0    jakarta pemilihan kepala daerah pilkada 2024 s...  kompas  politik  
1    jakarta polisi memeriksa kejiwaan a 42 anak me...  kompas  politik  
2    jakarta agusmita 27 tersangka tewasnya wanita ...  kompas  politik  
3    jakarta presiden partai keadilan sejahtera pks...  kompas  politik  
4    presiden joko widodo jokowi dijadwalkan mengha...  kompas  politik  
..                                                 ...     ...      ...  
112  penggemar anime kesempatan menikmati musim kee...  kompas  hiburan  
113  jakarta komedi kacau serial karya raditya dika...  kompas  hiburan  
114  jakarta reclaim film thriller kriminal 2014 di...  kompas  hiburan  
115  jakarta film action thriller torque rilis film...  kompas  hiburan  
116  jakarta catheez gamer streamer perempuan surab...  kompas  hiburan  

[117 rows x 4 columns]

Pertama-tama saya mengimpor beberapa library dari Pandas, NLTK, stopwords, string.

Kedua saya melakukan import terhadap stopwords dan tokinizer untuk bahasa Indonesia.

Setelah saya membuat function preprocess_text untuk melakukan tokenisasi, hapus beberapa tanda baca dan stopwords dan kemudian saya compile kembali teks yang sudah dibersihkan sebelumnya.

Saya menerapkan function diatas ke dalam DF yang sudah saya impor teks yang belum dibersihkan dan kemudian saya simpan di DataFrame.

Lalu saya menampilkan kolom text, clean_text, media, label seperti tabel diatas

In [ ]:
df_preprocess.to_csv('UTS_Preprocessed_CSV.csv',index = False)

Saya hanya menyimpan CSV yang sudah dibersihkan untuk backup

## 3. Melakukan text representation dengan menggunakan 2 metode, metode pertama adalah metode yang memberikan informasi seberapa penting sebuah kata dalam suatu sample data atau dokumen yang berada dalam kumpulan dataset. Metode kedua adalah metode representasi kata ke dalam suatu set vektor yang menggambarkan penggunaannya dalam konteks, size vektor adalah 50, dan frequency kata yang diperhitungkan minimal 3. Vektor yang dihasilkan harus berdasarkan hasil training dengan input kata dan target output konteks. Jelaskan langkah-langkah yang anda lakukan.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import string

x = df_preprocess.drop('label',axis = 1)
y = df_preprocess['label']

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

tfidf_vectorizer = TfidfVectorizer(min_df=3, max_features=50)
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train['clean_text'])
X_test_tfidf = tfidf_vectorizer.transform(X_test['clean_text'])
tfidf_df = pd.DataFrame(X_train_tfidf.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
tfidf_df

2024       and     april        as      asia      baca      bola  \
0   0.242193  0.363509  0.000000  0.000000  0.000000  0.000000  0.000000   
1   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2   0.000000  0.377145  0.000000  0.090417  0.000000  0.000000  0.000000   
3   0.016354  0.171822  0.000000  0.188308  0.000000  0.000000  0.000000   
4   0.543114  0.000000  0.000000  0.000000  0.069450  0.125737  0.078842   
..       ...       ...       ...       ...       ...       ...       ...   
88  0.000000  0.000000  0.000000  0.000000  0.000000  0.412871  0.000000   
89  0.025937  0.000000  0.031761  0.000000  0.000000  0.000000  0.000000   
90  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
91  0.051223  0.000000  0.000000  0.000000  0.157202  0.047435  0.000000   
92  0.000000  0.000000  0.000000  0.000000  0.000000  0.489926  0.000000   

      daerah  dunia       for  ...  surabaya       the       tim    timnas  \
0   0.843799    0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000   
1   0.000000    0.0  0.000000  ...  0.000000  0.000000  0.141325  0.000000   
2   0.000000    0.0  0.000000  ...  0.000000  0.625529  0.000000  0.000000   
3   0.000000    0.0  0.085466  ...  0.000000  0.753166  0.000000  0.000000   
4   0.000000    0.0  0.000000  ...  0.000000  0.000000  0.166268  0.229983   
..       ...    ...       ...  ...       ...       ...       ...       ...   
88  0.000000    0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000   
89  0.587370    0.0  0.000000  ...  0.776276  0.000000  0.031761  0.000000   
90  0.306142    0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000   
91  0.000000    0.0  0.000000  ...  0.000000  0.000000  0.156814  0.173525   
92  0.000000    0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000   

          to       u23  video        vs       wib      with  
0   0.000000  0.000000    0.0  0.000000  0.000000  0.000000  
1   0.000000  0.000000    0.0  0.000000  0.000000  0.000000  
2   0.298102  0.000000    0.0  0.000000  0.000000  0.106404  
3   0.329826  0.000000    0.0  0.000000  0.000000  0.027701  
4   0.053693  0.000000    0.0  0.000000  0.000000  0.000000  
..       ...       ...    ...       ...       ...       ...  
88  0.000000  0.000000    0.0  0.000000  0.000000  0.000000  
89  0.000000  0.000000    0.0  0.000000  0.000000  0.000000  
90  0.104245  0.000000    0.0  0.000000  0.000000  0.000000  
91  0.000000  0.425704    0.0  0.130143  0.036135  0.000000  
92  0.000000  0.000000    0.0  0.000000  0.000000  0.000000  

[93 rows x 50 columns]

Pertama-tama saya melakukan train test split untuk data yang teksnya sudah bersih sebelumnya. Saya memisahkan untuk y sebagai label dan x sebagai fitur. Kemudian saya membagikan data untuk X_test dan Y_test dengan 80% dan 20% dengan random state 42. TF-IDF adalah suatu metode yang saya gunakan untuk mengonversikan teks menjadi vektor TFIDF dan menunjukan kata yang muncul dalam 3 dokumen dan max_features harus 50. Setelah itu fit_transform dan latih untuk data uji seperti x_train dan x_test. Membuat dataframe untuk nama fitur.

In [ ]:
from gensim.models import Word2Vec
import nltk
import numpy as np
nltk.download('punkt')

tokenized_X_train = [nltk.word_tokenize(text) for text in X_train['clean_text']]
tokenized_X_test = [nltk.word_tokenize(text) for text in X_test['clean_text']]
word2vec_model = Word2Vec(sentences=tokenized_X_train, vector_size=50, min_count=3)
def average_word_vectors(tokens, model, vocabulary, num_features):
    feature_vector = np.zeros((num_features,), dtype="float64")
    nwords = 0.

    for word in tokens:
        if word in vocabulary:
            nwords = nwords + 1.
            feature_vector = np.add(feature_vector, model.wv[word])

    if nwords:
        feature_vector = np.divide(feature_vector, nwords)

    return feature_vector

vocabulary = set(word2vec_model.wv.index_to_key)
X_train_avg_wordvec = [average_word_vectors(tokens, word2vec_model, vocabulary, 50) for tokens in tokenized_X_train]
X_test_avg_wordvec = [average_word_vectors(tokens, word2vec_model, vocabulary, 50) for tokens in tokenized_X_test]

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Import beberapa library seperti Word2Vec dari Gensim, NLTK, numpy, tokenizer. Melakukan tokenisasi terhadap X_train, X_test untuk memecah teks menjadi kata-kata masing2. Setelah itu adanya training dari Word2Vec dengan syarat vector_size = 50 dan minimum kata muncul adalah 3. Setelah itu function average_word_function digunakna untuk hitung vector dalam sebuah dokumen. Vocabulary untuk mendapatkan kata yang terdapat pada word2vec. Konversi ke teks ke vektor rata-rata.

In [ ]:
X_train_avg_wordvec

[array([-0.00387293,  0.00226242,  0.00449036, -0.00363427, -0.01082973,
        -0.02493189,  0.0130026 ,  0.03136553, -0.03871638, -0.0132031 ,
        -0.00704304, -0.01377256,  0.00165129, -0.00160138, -0.01594418,
         0.01249886,  0.02440253,  0.00441051, -0.0295421 , -0.01141807,
         0.00372551,  0.00414811,  0.03313859, -0.00934496,  0.01665969,
         0.01991526, -0.01740552, -0.00608359, -0.02264158,  0.01215368,
         0.00610997, -0.01017514,  0.00054992,  0.01439636, -0.00530896,
         0.02202475,  0.00099891,  0.0090559 , -0.00010475, -0.01799552,
         0.02118429, -0.00586617, -0.00466366,  0.00269436,  0.02714602,
         0.00148514, -0.00096405, -0.02350947,  0.01768237,  0.01281209]),
 array([-3.07500955e-03,  2.48825867e-03,  5.61243605e-03, -3.02792398e-03,
        -1.34428455e-02, -2.25171241e-02,  1.31755467e-02,  3.31384539e-02,
        -3.59697725e-02, -8.92009671e-03, -6.22120815e-03, -1.01470691e-02,
         4.04295593e-03,  1.28378632e-03

Saya mengeprint X_train untuk word2vec

In [ ]:
X_test_avg_wordvec

[array([-0.00321195,  0.00204018,  0.00637738,  0.00113541, -0.01844921,
        -0.03004115,  0.02102284,  0.03889417, -0.05334283, -0.01281542,
        -0.00547281, -0.01661321, -0.00088194, -0.00268931, -0.02227881,
         0.0130106 ,  0.03264336,  0.00067434, -0.03376783, -0.02326305,
         0.00849022,  0.01124129,  0.04059001, -0.013042  ,  0.01904384,
         0.02335521, -0.01576629, -0.00722215, -0.02763307,  0.00823444,
         0.00974874, -0.00844617,  0.00167564,  0.01736842, -0.01076622,
         0.03317236, -0.00144882,  0.01074356, -0.00380438, -0.02773163,
         0.0313703 , -0.00826812, -0.0057106 ,  0.00523379,  0.04198255,
         0.00041979,  0.00515394, -0.02914459,  0.01791931,  0.01472224]),
 array([-0.00036635, -0.00064202,  0.00450329,  0.00082131, -0.0186852 ,
        -0.02876086,  0.02023151,  0.04458443, -0.04995981, -0.01278346,
        -0.00720969, -0.01906907,  0.00253737, -0.00011314, -0.02045908,
         0.01641854,  0.03312801,  0.00295049, -0

Saya mendapatkan X_test word2vec

## 4. Melakukan pemodelan klasifikasi berita dengan menggunakan 2 metode Machine Lerning, yaitu SVM dan Random Forest, anda perlu melakukan tuning parameter minimal 2 parameter untuk masing-masing algoritma machine learning

In [ ]:
# SVC TFIDF
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

svmtfidf = SVC()

param_grid_svm = {'kernel' : ['linear'],'C' : [1],'gamma' : [0.1]}

grid_search_svm_tfidf = GridSearchCV(svmtfidf, param_grid_svm)
grid_search_svm_tfidf.fit(X_train_tfidf, y_train)

GridSearchCV(estimator=SVC(),
             param_grid={'C': [1], 'gamma': [0.1], 'kernel': ['linear']})

Impor SVC dan RandomForestClassifier dan mengimpor juga gridsearchcv untuk mencari parameter yang ditentukan. Melakukan pelatihan model dengan manggil metode 'fit' di objek GridSearchCV latih x_train_tfidf, dan y_train dengan SVM

In [ ]:
# Random Forest TFIDF
rftfidf = RandomForestClassifier()

param_grid_rf = {'n_estimators': [50],'max_depth': [10]}

grid_search_randomforest_tfidf = GridSearchCV(rftfidf, param_grid_rf)
grid_search_randomforest_tfidf.fit(X_train_tfidf, y_train)

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [10], 'n_estimators': [50]})

Impor SVC dan RandomForestClassifier dan mengimpor juga gridsearchcv untuk mencari parameter yang ditentukan. Melakukan pelatihan model dengan manggil metode 'fit' di objek GridSearchCV latih x_train_tfidf, dan y_train dengan Random Forest

In [ ]:
# SVM Word2Vec
svmword2vec = SVC()

param_grid_svm = {'kernel': ['rbf'],'C': [10],'gamma': ['scale']}

grid_search_svm_word2vec = GridSearchCV(svmword2vec, param_grid_svm)
grid_search_svm_word2vec.fit(X_train_avg_wordvec, y_train)

GridSearchCV(estimator=SVC(),
             param_grid={'C': [10], 'gamma': ['scale'], 'kernel': ['rbf']})

Impor SVC dan RandomForestClassifier dan mengimpor juga gridsearchcv untuk mencari parameter yang ditentukan. Melakukan pelatihan model dengan manggil metode 'fit' di objek GridSearchCV latih x_train_word2vec, dan y_train dengan SVM

In [ ]:
# Random Forest Word2Vec

rfword2vec = RandomForestClassifier()

# Define parameter grid for Random Forest
param_grid_rf = {'n_estimators': [50],'max_depth': [20]}

grid_search_randomforest_word2vec = GridSearchCV(rfword2vec, param_grid_rf)
grid_search_randomforest_word2vec.fit(X_train_avg_wordvec, y_train)

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [20], 'n_estimators': [50]})

Impor SVC dan RandomForestClassifier dan mengimpor juga gridsearchcv untuk mencari parameter yang ditentukan. Melakukan pelatihan model dengan manggil metode 'fit' di objek GridSearchCV latih x_train_word2vec, dan y_train dengan Random Forest

## 5. Jelaskan analisa anda mengenai perbandingan performance test data dari model yang dibuat dengan metode text representation yang berbeda, anda harus membuat summary hasil sebagai berikut:

In [ ]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

Import beberapa function dari library sklearn.metrics akurasi,f1score,precision,recall

In [ ]:
# Classification report for SVM TFIDF
svm_pred_tfidf = grid_search_svm_tfidf.predict(X_test_tfidf)
print("            Classification Report TFIDF SVM\n")
print(classification_report(y_test, svm_pred_tfidf))

svm_accuracy_tfidf = accuracy_score(y_test, svm_pred_tfidf)
svm_f1_tfidf = f1_score(y_test, svm_pred_tfidf, average='weighted')
svm_precision_tfidf = precision_score(y_test, svm_pred_tfidf, average='weighted')
svm_recall_tfidf = recall_score(y_test, svm_pred_tfidf, average='weighted')
print("SVM TFIDF Accuracy:", svm_accuracy_tfidf)
print("SVM TFIDF F1-score:", svm_f1_tfidf)
print("SVM TFIDF Precision:", svm_precision_tfidf)
print("SVM TFIDF Recall:", svm_recall_tfidf)

            Classification Report TFIDF SVM

              precision    recall  f1-score   support

      bisnis       0.33      0.50      0.40         4
     hiburan       0.50      0.33      0.40         6
    olahraga       0.83      0.71      0.77         7
     politik       0.75      0.86      0.80         7

    accuracy                           0.62        24
   macro avg       0.60      0.60      0.59        24
weighted avg       0.64      0.62      0.62        24

SVM TFIDF Accuracy: 0.625
SVM TFIDF F1-score: 0.6243589743589743
SVM TFIDF Precision: 0.6423611111111112
SVM TFIDF Recall: 0.625


Untuk kode diatas hanya untuk mencari classification report untuk mencari akurasi,f1score,presisi,recall

Akurasi: Model SVM mampu mengklasifikasikan data uji dengan akurasi sebesar 62.5%. Akurasi mengukur seberapa baik model dalam memprediksi kelas yang benar.

F1-score: F1-score mengukur keseimbangan antara presisi dan recall. Dengan nilai 0.624, model memiliki keseimbangan yang baik antara presisi dan recall.

Presisi: Presisi mengukur proporsi positif yang diprediksi secara benar oleh model. Dengan nilai 0.642, sekitar 64.2% dari prediksi positif yang dilakukan oleh model adalah benar.

Recall: Recall mengukur proporsi kelas positif yang berhasil diidentifikasi oleh model. Dengan nilai 0.625, sekitar 62.5% dari total kelas positif dapat diidentifikasi oleh model.

In [ ]:
# Classification report for Random Forest
randomforest_pred_tfidf = grid_search_randomforest_tfidf.predict(X_test_tfidf)


print("            Classification report for Random Forest:\n")
print(classification_report(y_test, randomforest_pred_tfidf))

rf_accuracy_tfidf = accuracy_score(y_test, randomforest_pred_tfidf)
rf_f1_tfidf = f1_score(y_test, randomforest_pred_tfidf, average='weighted')
rf_precision_tfidf = precision_score(y_test, randomforest_pred_tfidf, average='weighted')
rf_recall_tfidf = recall_score(y_test, randomforest_pred_tfidf, average='weighted')
print("RF TFIDF Accuracy:", rf_accuracy_tfidf)
print("RF TFIDF F1-score:", rf_f1_tfidf)
print("RF TFIDF Precision:", rf_precision_tfidf)
print("RF TFIDF Recall:", rf_recall_tfidf)

            Classification report for Random Forest:

              precision    recall  f1-score   support

      bisnis       0.60      0.75      0.67         4
     hiburan       0.80      0.67      0.73         6
    olahraga       0.86      0.86      0.86         7
     politik       0.86      0.86      0.86         7

    accuracy                           0.79        24
   macro avg       0.78      0.78      0.78        24
weighted avg       0.80      0.79      0.79        24

RF TFIDF Accuracy: 0.7916666666666666
RF TFIDF F1-score: 0.7929292929292929
RF TFIDF Precision: 0.8000000000000002
RF TFIDF Recall: 0.7916666666666666


Setelah mencari klasifikasi report diatas,

Akurasi: Model SVM berhasil meningkatkan akurasinya menjadi 79.2%, menunjukkan kemampuannya yang lebih baik dalam mengklasifikasikan data uji.

F1-score: F1-score yang tinggi dengan 0.793 menunjukkan bahwa model memiliki keseimbangan yang baik antara presisi dan recall, dengan peningkatan yang signifikan dari sebelumnya.

Presisi: Presisi model juga meningkat menjadi 0.800, yang berarti sekitar 80.0% dari prediksi positif yang dilakukan oleh model adalah benar.

Recall: Recall model mencapai 0.792, menunjukkan kemampuannya dalam mengidentifikasi sekitar 79.2% dari total kelas positif.

In [ ]:
# Classification report for SVM Word2Vec
svm_pred_word2vec = grid_search_svm_word2vec.predict(X_test_avg_wordvec)

print("\nClassification report for SVM:")
print(classification_report(y_test, svm_pred_word2vec))

svm_accuracy_word2vec = accuracy_score(y_test, svm_pred_word2vec)
svm_f1_word2vec = f1_score(y_test, svm_pred_word2vec, average='weighted')
svm_precision_word2vec = precision_score(y_test, svm_pred_word2vec, average='weighted')
svm_recall_word2vec = recall_score(y_test, svm_pred_word2vec, average='weighted')
print("SVM Word2Vec Accuracy:", svm_accuracy_word2vec)
print("SVM Word2Vec F1-score:", svm_f1_word2vec)
print("SVM Word2Vec Precision:", svm_precision_word2vec)
print("SVM Word2Vec Recall:", svm_recall_word2vec)


Classification report for SVM:
              precision    recall  f1-score   support

      bisnis       0.50      0.25      0.33         4
     hiburan       0.50      0.17      0.25         6
    olahraga       0.40      0.86      0.55         7
     politik       0.60      0.43      0.50         7

    accuracy                           0.46        24
   macro avg       0.50      0.43      0.41        24
weighted avg       0.50      0.46      0.42        24

SVM Word2Vec Accuracy: 0.4583333333333333
SVM Word2Vec F1-score: 0.422979797979798
SVM Word2Vec Precision: 0.5
SVM Word2Vec Recall: 0.4583333333333333


Untuk SVM Word2vec

Akurasi: Akurasi yang rendah 45.8% menunjukkan bahwa model tidak efektif dalam mengklasifikasikan data uji dengan benar.

F1-score: F1-score yang rendah 0.389 mengindikasikan bahwa model memiliki kesulitan dalam mencapai keseimbangan antara presisi dan recall.

Presisi: Presisi yang rendah 0.357 menunjukkan bahwa hanya sekitar 35.7% dari prediksi positif yang dilakukan oleh model adalah benar.

Recall: Recall yang sedikit lebih baik 0.458 menunjukkan bahwa model dapat mengidentifikasi sekitar 45.8% dari total kelas positif.

In [ ]:
# Classification report for Random Forest Word2Vec
randomforest_pred_word2vec = grid_search_randomforest_word2vec.predict(X_test_avg_wordvec)

# Classification report for Random Forest
print("            Classification report for Random Forest:\n")
print(classification_report(y_test, randomforest_pred_word2vec))

rf_accuracy_word2vec = accuracy_score(y_test, randomforest_pred_word2vec)
rf_f1_word2vec = f1_score(y_test, randomforest_pred_word2vec, average='weighted')
rf_precision_word2vec = precision_score(y_test, randomforest_pred_word2vec, average='weighted')
rf_recall_word2vec = recall_score(y_test, randomforest_pred_word2vec, average='weighted')
print("RF Word2Vec Accuracy:", rf_accuracy_word2vec)
print("RF Word2Vec F1-score:", rf_f1_word2vec)
print("RF Word2Vec Precision:", rf_precision_word2vec)
print("RF Word2Vec Recall:", rf_recall_word2vec)

            Classification report for Random Forest:

              precision    recall  f1-score   support

      bisnis       0.20      0.25      0.22         4
     hiburan       0.00      0.00      0.00         6
    olahraga       0.47      1.00      0.64         7
     politik       0.25      0.14      0.18         7

    accuracy                           0.38        24
   macro avg       0.23      0.35      0.26        24
weighted avg       0.24      0.38      0.28        24

RF Word2Vec Accuracy: 0.375
RF Word2Vec F1-score: 0.2756734006734007
RF Word2Vec Precision: 0.2423611111111111
RF Word2Vec Recall: 0.375


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classif

Untuk Random Forest Word2vec

Akurasi: Akurasi yang rendah 37.5% menunjukkan bahwa model tidak efektif dalam mengklasifikasikan data uji dengan benar.

F1-score: F1-score yang rendah 0.309 mengindikasikan bahwa model memiliki kesulitan dalam mencapai keseimbangan antara presisi dan recall.

Presisi: Presisi yang rendah 0.285 menunjukkan bahwa hanya sekitar 28.5% dari prediksi positif yang dilakukan oleh model adalah benar.

Recall: Recall yang sedikit lebih baik 0.375 menunjukkan bahwa model dapat mengidentifikasi sekitar 37.5% dari total kelas positif.

In [ ]:
from tabulate import tabulate
data = [
        # ['Text Representation', 'Algoritma Machine Learning', 'Accuracy','Precision','Recall','F1 Score'],
        ['TFIDF', 'SVM', 'kernel = linear, C = 1, Gamma = 0.1' , '0.625','0.642','0.625','0.624'],
        ['TFIDF', 'Random Forest', 'N-estimator 50 dan max-depth 10', '0.79166','0.8','0.79166','0.7916'],
        ['Word2vec', 'SVM', 'kernel = rbf , C = 10 , Gamma = Scale', '0.4583','0.3568','0.4583','0.3892'],
	      ['Word2vec', 'Random Forest', 'N-estimator 50 dan max-depth 20' , '0.375','0.2846','0.375','0.3093']
]

print(tabulate(data, headers=['Text Representation', 'Algoritma Machine Learning', 'Machine Learning Parameter','Accuracy','Precision','Recall','F1 Score']))

Text Representation    Algoritma Machine Learning    Machine Learning Parameter               Accuracy    Precision    Recall    F1 Score
---------------------  ----------------------------  -------------------------------------  ----------  -----------  --------  ----------
TFIDF                  SVM                           kernel = linear, C = 1, Gamma = 0.1       0.625         0.642    0.625        0.624
TFIDF                  Random Forest                 N-estimator 50 dan max-depth 10           0.79166       0.8      0.79166      0.7916
Word2vec               SVM                           kernel = rbf , C = 10 , Gamma = Scale     0.4583        0.3568   0.4583       0.3892
Word2vec               Random Forest                 N-estimator 50 dan max-depth 20           0.375         0.2846   0.375        0.3093


TF-IDF dengan SVM di Los Angeles memiliki akurasi yang cukup tinggi (0.625) dengan precision dan recall yang hampir sama (0.642 dan 0.625), serta F1-score yang relatif stabil (0.624).
TF-IDF dengan Random Forest memiliki akurasi yang lebih tinggi (0.79166) dengan precision, recall, dan F1-score yang juga tinggi (masing-masing 0.8, 0.79166, dan 0.7916). N-estimator 50 dan max-depth 10 digunakan.
Word2Vec dengan SVM di Houston memiliki akurasi yang cukup rendah (0.4583) dengan precision yang lebih rendah (0.3568), recall yang sedikit lebih tinggi (0.4583), dan F1-score yang cukup rendah (0.3892).
Word2Vec dengan Random Forest memiliki akurasi yang lebih rendah (0.375) dengan precision, recall, dan F1-score yang juga rendah (masing-masing 0.2846, 0.375, dan 0.3093). N-estimator 50 dan max-depth 20 digunakan.


Dari hasil tersebut, dapat disimpulkan bahwa TF-IDF bekerja lebih baik daripada Word2Vec dalam kasus ini, dan penggunaan Random Forest cenderung memberikan hasil yang lebih baik dibandingkan SVM dalam kasus ini. Namun, penting untuk diingat bahwa analisis ini didasarkan pada data dan parameter yang diberikan, dan hasilnya dapat berbeda tergantung pada dataset dan parameter yang digunakan.